src/Testcontainers/Builders/TlsEndpointAuthenticationProvider.cs

```diff
@@ -0,0 +1,105 @@
+namespace DotNet.Testcontainers.Builders
+{
+  using System;
+  using System.IO;
+  using System.Linq;
+  using System.Net.Security;
+  using System.Security.Cryptography.X509Certificates;
+  using DotNet.Testcontainers.Configurations;
+
+  /// <inheritdoc cref="IDockerRegistryAuthenticationProvider" />
+  internal class TlsEndpointAuthenticationProvider : DockerEndpointAuthenticationProvider
+  {
+    protected const string CaCertificateFileName = "ca.pem";
+
+    protected const string ClientCertificateFileName = "cert.pem";
+
+    protected const string ClientCertificateKeyFileName = "key.pem";
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="TlsEndpointAuthenticationProvider" /> class.
+    /// </summary>
+    /// <param name="customConfigurations">A list of custom configurations.</param>
+    public TlsEndpointAuthenticationProvider(params ICustomConfiguration[] customConfigurations)
+      : this(customConfigurations
+        .OrderByDescending(item => item.GetDockerTls() && item.GetDockerHost() != null)
```

Comment:
Why do you need ordering based on Tls. This does not reflect TlsVerify which is need just for mTls. I think it goes little bit  against. Tls is set to true and TlsVerify can be false. Then TlsVerify can be true but Tls can be false, but it does not matter becuase client take it as true of of the box.

Reply:
Also I think that we should never change order and follow overriding rule which is ,I guess, Environment variable ->Testcontainer setting file -> (maybe in future) command line arguments -> settings set by code -> finall settings.
Probably for now it is ok, not sure if tlsVerify will not be affected then.

Reply:
Yep, you are right, I'll update it tomorrow.

Reply:
I think the order does not matter here (at least for now). If `GetDockerTls()` or `GetDockerTlsVerify()` is not set, it is always false. Hence, we cannot override it. We pick the first configuration that enables it. To support to override the properties we need them nullable. Same for

https://github.com/testcontainers/testcontainers-dotnet/blob/78cf9a3b910ef73088142ebcc267d8127b54f68f/src/Testcontainers/Configurations/TestcontainersSettings.cs#L40

We can add that feature in a new PR (we need to switch the order in `TestcontainersSettings` too). But you are right, the ordering above is wrong.

Reply:
I think we can leave it after that release. and do more regarding settings.